src/verifast1.ml

```diff
@@ -4237,10 +4373,23 @@ module VerifyProgram1(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
     | InstPredAsn (l, e, g, index, pats) ->
       let (w, t) = check_expr (pn,ilist) tparams tenv (Some true) e in
       begin match unfold_inferred_type t with
-        ObjType cn ->
-        let check_call family pmap =
-          let (wpats, tenv) = check_pats (pn,ilist) l tparams tenv (List.map snd pmap) pats in
-          let index = check_expr_t (pn,ilist) tparams tenv (Some true) index (ObjType "java.lang.Class") in
+        ObjType (cn, targs) ->
+        let check_call family pmap tparams =
+          let ptps = List.map snd pmap in
+          let tpenv = 
+            if (List.length targs > 0) then (if (List.length tparams) <> (List.length targs) 
+            then static_error l (Printf.sprintf "Wrong amount of type arguments provided. Expected: %s Actual %s." (string_of_int (List.length tparams)) (string_of_int (List.length targs))) None
+            else List.map2 (fun a b -> (a,b)) tparams targs) else [] in
+          (* We know both type params and type arguments here, if possible fill them in *)
+          let rec replaceTypeParams t = match t with 
+            GhostTypeParam t -> (try List.assoc t tpenv with _ -> GhostTypeParam t)
+            | ObjType (s,ts) -> ObjType (s, List.map replaceTypeParams ts)
+            | InductiveType (s,ts) -> InductiveType (s, List.map replaceTypeParams ts)
+            | PredType (s,ts,o,i) -> PredType (s, List.map replaceTypeParams ts, o, i)
+            | PureFuncType (a,b) -> PureFuncType (replaceTypeParams a, replaceTypeParams b)
+            | t -> t  in
+          let (wpats, tenv) = check_pats (pn,ilist) l tparams tenv (List.map replaceTypeParams ptps) pats in
+          let index = check_expr_t (pn,ilist) tparams tenv (Some true) index (ObjType ("java.lang.Class", [])) in

```

Comment:
Ik denk dat dit unsound is. Beschouw:
```java
class A {
  //@ predicate foo() = true;
  
  A()
    //@ requires true;
    //@ ensures foo();
  {}
}

class B extends A {
  //@ predicate foo() = false;
  
  /*@
  lemma void foo_is_absurd()
    requires foo();
    ensures false;
  {
    open foo();
  }
  @*/
}

class Cell<T> {
  T contents;
  //@ predicate Cell(T contents) = this.contents |-> contents;
  
  Cell(T contents)
    //@ requires true;
    //@ ensures Cell(contents);
  {
    this.contents = contents;
  }
}

class Test {
  static void main()
    //@ requires true;
    //@ ensures false;
  {
    Cell<A> cell = new Cell<A>(new A());
    Cell<B> cellB = cell;
    //@ assert cellB.Cell(?contents);
    //@ contents.foo_is_absurd();
  }
}
```
Aangezien je geen voorkomens van typeparameters toelaat in instance predicate definitions, is dit sowieso dode code? Dus liefst verwijderen.

Reply:
Ik denk dat dit inderdaad nog oude code is van wanneer Java type parameters nog konden terug komen in ghost code.